Publicize: Force theme support for post-thumbnails feature

p-jackson opened the pull request:
Forces the "Featured image" UI to appear in the block editor when the publicize module is enabled, even if the theme doesn't support the "post-thumbnails" feature.

Posts to Twitter and Facebook use the featured image if it's available, so even if a theme doesn't support it there's still value in allowing users to set it.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

_Before_
<img width="1598" alt="Screenshot 2022-04-08 at 3 28 44 PM" src="https://user-images.githubusercontent.com/1500769/162358852-5837e0bb-01c9-4cb0-86e7-66ef0e7d7261.png">

_After_
<img width="1598" alt="Screenshot 2022-04-08 at 3 24 17 PM" src="https://user-images.githubusercontent.com/1500769/162358828-9d08ed8b-25cf-40fd-92aa-6087e9858cc0.png">

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Add `'post-thumbnails'` support to the current theme, which causes the featured image UI to appear in the block editor sidebar.
* Add new `'jetpack_publicize_force_featured_image_panel'` filter which can be used to disable this behaviour.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
Fixes Automattic/wp-calypso#55081 on WordPress.com, but this should really be a fix for all Publicize users p1HpG7-frH-p2


#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Install a theme that doesn't support the `post-thumbnails` feature ([you can download Plane from WordPress.com](https://public-api.wordpress.com/rest/v1/themes/download/plane.zip))
* Ensure Publicize module is enabled
* Open post editor
   * The "Featured image" UI should be available in the sidebar
   * Confirm the network tools confirm the `/wp/v2/themes` API response includes `post-thumbnails=true`
* Disable the Publicize module
* Reopen post editor
   * The "Featured image" UI shouldn't be in the sidebar
   * Confirm the network tools confirm the `/wp/v2/themes` API response includes `post-thumbnails=false`

When I connected to Twitter and tested publicize I didn't see the featured image actually show up in the tweet. I put this down to a limitation in `jurassic.tunnel` maybe? When I do the same on WordPress.com the image appears in the tweet.

p-jackson committed:
[not verified] Publicize: Force theme support for post-thumbnails feature

Forces the "Featured image" UI to appear in the block editor when the
publicize module is enabled, even if the theme doesn't support the
"post-thumbnails" feature.

Posts to Twitter and Facebook use the featured image if it's available,
so even if a theme doesn't support it there's still value in allowing
users to set it.

p-jackson committed:
[not verified] changelog

p-jackson commented:
> * [ ] Publish a post with Facebook integration

I'm having trouble connecting to my test Facebook account 😕 

<img width="674" alt="Screenshot 2022-04-08 at 5 42 09 PM" src="https://user-images.githubusercontent.com/1500769/162372156-498c7f6a-6442-45c9-bc16-149e2a24edda.png">

jeherve reviewed (commented):
Thanks for the PR. Two things come to my mind:

1. Would it be possible to add this to `modules/publicize/publicize.php` instead? The base class is synchronized with WordPress.com, so WordPress.com would benefit from your changes at the same time.
2. Would it be an option to follow the existing pattern used there, where we hook into `init` to add support for a feature (here post-thumbnails)?
https://github.com/Automattic/jetpack/blob/02efbf941065d74d0049945159b95906a4895ce7/projects/plugins/jetpack/modules/publicize/publicize.php#L220
In this case, we would be adding support for `post-thumbnails`, maybe just with something like this, hooked into `init`, to keep things simple?

```php
add_theme_support( 'post-thumbnails' );
```

Would that be enough? Would it work?

p-jackson commented:
> Would it be possible to add this to `modules/publicize/publicize.php` instead?

👍 Will do

As for using `add_theme_support( 'post-thumbnails' );` yes I've tried this and it does work. I used the filter instead because that's the tried and true method that's been working on wpcom, so we know it doesn't cause problems. But simply adding theme support seems more straightforward and means internal APIs will be consistent with the REST API response. So perhaps it is better just to use this method.

danielpost commented:
@p-jackson Thanks for the improvement! This tests well for me. It doesn't work in the Classic Editor, which makes sense since that won't use the REST response. I suspect using @jeherve's approach would add support for the Classic Editor as well.

In addition to Jeremy's suggestion, I think we should also make this filterable in some way. Considering the Open Graph images are filterable as well, I can imagine some scenarios where you don't really want a Featured Image panel to appear. Something like `apply_filters( 'jetpack_publicize_force_featured_image_panel', true )` perhaps?

(thinking about this a little more, I'm wondering if we should even keep the default to false, thus not changing anything but adding the option for users that need it...)

p-jackson commented:
Thanks for the review @danielpost, that's a very good point about the classic editor and a good reason to switch to `add_theme_support()`.

I've been extra polite and only added support for the post type that we're interested in. `add_theme_support()` merges the post type parameter so we don't need to worry about clobbering existing settings.

> I'm wondering if we should even keep the default to false, thus not changing anything but adding the option for users that need it...

This is a fair product question. I'm not as familiar with what Jetpack customers would want, and @jeherve is the only person who's weighed in so far on p1HpG7-frH-p2. I'm inclined to think that an opt-out of the new behaviour is best (the featured image accordion isn't expanded by default so not very intrusive), but happy to default to false in Jetpack.

pablinos commented:
> I'm wondering if we should even keep the default to false, thus not changing anything but adding the option for users that need it...

Having it default to `false` might not be practical, as turning it on would require access to the code, and then support could be added to the theme anyway, in the same way as we're talking about here.

An option could be to add theme support on `init`, as @jeherve suggests, but at a higher priority than 10, which would mean any existing code, hooked onto `init` to remove thumbnail support should still function correctly.

pablinos commented on projects/plugins/jetpack/modules/publicize/publicize.php:
```diff
@@ -220,6 +220,10 @@ public function __construct() {
 		add_action( 'init', array( $this, 'register_post_meta' ), 20 );
 		add_action( 'jetpack_register_gutenberg_extensions', array( $this, 'register_gutenberg_extension' ) );
+
+		// The custom priority for this action allows a 'jetpack_publicize_force_featured_image_panel'
```
As mentioned in my comment, I wonder if this prority should go the other way, say 8, to ensure that any code to remove support for thumbnails from a theme would still have an opportunity to run.

It might also mean that we don't need the `jetpack_publicize_force_featured_image_panel` because someone could call `remove_action` on `add_theme_post_thumbnails_support`.

p-jackson replied:
I think I understand the rationale. It means that we wouldn't break existing code that already calls `remove_theme_support( 'post-thumbnails' )`.

> It might also mean that we don't need the `jetpack_publicize_force_featured_image_panel` because someone could call `remove_action` on `add_theme_post_thumbnails_support`.

How easy is it to call `remove_action` when the callback is a member function? That might be some PHP foo I'm unaware of, but it seems to me that `remove_action` only works when the callback is a global function?

pablinos replied:
Yes, it can be done, but it's difficult without a global reference. There is a global `$publicize` variable that should be the object we'd need, so we could document that, or we could make the function static, as it's not relying on anything else in the class.

Alternatively, all [this code is going to live in a namespace now](https://github.com/Automattic/jetpack/blob/master/projects/packages/publicize/src/class-publicize-base.php#L10), so we could have the function outside of the class instead. We are currently in the process of getting Jetpack to use this package instead, so if we go this route we could maybe wait for that, and adapt this change then.

p-jackson commented:
Thanks for the feedback @pablinos, I've updated the PR to include it. It makes sense to add our theme support early (in a higher priority hook) so that any theme or plugin code that was removing featured image support will continue to work. I've also moved the hook callback to a non-member function to make it easier to remove the action handler at runtime, with the expectation that this file will be moving inside a namespace soon and I'll be able to rebase this PR when that happens. Do you have an idea of the timeline for that?

Also, do you have an idea about why the E2E tests are failing for this PR? It looks like it's a sync test that's failing, but I can't see how changing support for a theme feature would be breaking synchronisation 🤔

pablinos reviewed (dismissed):
Sorry @p-jackson, I meant to respond to this, and then Easter got in the way. This looks good to me, but we'll need to merge master (or rebase) to get it up to date.

> Also, do you have an idea about why the E2E tests are failing for this PR? 

No, I think that test has been failing on other PRs and can probably be ignored. The rebase might help here too, but I think as these tests can fail from time to time, it's one of the reasons that the check isn't required.

pablinos commented:
@Automattic/jetpack-reach Just noting that once this is merged, we'll need to migrate the change to the package.

jeherve commented on projects/plugins/jetpack/modules/publicize/publicize.php:
```diff
@@ -1456,3 +1460,13 @@ protected static function build_sprintf( $args ) {
+ * explicitly set an image for their social media post.
+ */
+function add_theme_post_thumbnails_support() {
+	add_theme_support( 'post-thumbnails', array( 'post' ) );
```
Instead of only adding it to the post editor, should we aim to add support to any post type that supports Publicize?

See this for prior art on this:
https://github.com/Automattic/jetpack/blob/60ff8e8f3400ab8e8bf7c68f1377e506d6f1cfd9/projects/plugins/jetpack/_inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php#L66-L67

p-jackson replied:
Thanks! This was my intention but I thought that only `'post'` types supported Publicize. `get_post_types_by_support()` is actually what I want 👍

p-jackson replied:
I've changed the hardcoded array to `get_post_types_by_support( 'publicize' )`, however I've also had to change the priority of the hook that adds `'publicize'` support to the `'post'` post type since they both happen during the `init` action.

https://github.com/Automattic/jetpack/blob/958e0b05e930e8df6e90c3f4ef61b9b6ba0836b9/projects/packages/publicize/src/class-publicize-base.php#L222

coder-karen commented:
@p-jackson heads up that the code freeze for the Jetpack 10.9 milestone is today, just in case the intention is for this PR to be included in the 10.9 release. Let us know (Crew) if there is anything we can do to help if so!

p-jackson commented:
Thanks for the heads up @coder-karen, I was similarly interrupted by Easter. Due to timezones this probably won't make it into `10.9`. We can probably ship a temporary fix (884-gh-Automattic/wpcomsh) which will at least help WPCOM users while we wait for version `10.10`

coder-karen commented:
Sounds good @p-jackson . Technically we have about 5hrs before cutting the beta if that helps at all, otherwise the temporary fix sounds like a good idea.

p-jackson committed:
[not verified] Support all publicize post types and move code to package

p-jackson committed:
[not verified] changelog

p-jackson commented:
I've [incorporated the feedback above](https://github.com/Automattic/jetpack/pull/23871#discussion_r856357451) and the `post-thumbnails` feature will now be enabled for all supported post types, not just the default `'post'`. However, things have changed with Publicize recently so it's not working exactly right. #23786 changed Publicize to use the composer package instead of the module. Afaict the code in the module directory is no longer loaded so I've moved my changes from `/plugins/jetpack/modules/publicize/publicize.php` over to `/packages/publicize/src/class-publicize-base.php`. I'm also now able to use a namespace for the action callback.

However, previously the `Publicize_Base` class wasn't instantiated when the Publicize feature was disabled in the settings. But the `Publicize_Base` class in the package _is_ instantiated when the Publicize feature is disabled. That means even when Publicize is toggled off, the featured image UI is appearing in the editor.

I think this is the line that means it's always loaded:
https://github.com/Automattic/jetpack/blob/49de89141b150cd9107fb7633f05df41318adda0/projects/plugins/jetpack/class.jetpack.php#L827

@jeherve have you got any guidance on the correct approach to detecting when Publicize is enabled from inside `Publicize_Base`?

spsiddarthan commented:
> @jeherve have you got any guidance on the correct approach to detecting when Publicize is enabled from inside Publicize_Base?I

I'll let @jeherve confirm if this is the right approach, but I wonder if we should ensure Publicize is loaded from within Jetpack only if the publicize module is active. 

```
use Automattic\Jetpack\Modules;

if (( new Modules() )->is_active( 'publicize' )) {
			$config->ensure( 'publicize' );
}
```

spsiddarthan commented:
I tried implementing my solution above, but when I go to the modules screen and deactivate publicize and try activiating publicize again, I got this strange error. 

<img width="1698" alt="Screenshot 2022-05-03 at 1 28 31 PM" src="https://user-images.githubusercontent.com/6594561/166420933-0bcdce30-c1ac-47bd-9346-9e20b53fd187.png">

Maybe as a stop-gap, we can add the check to see if the publicize module is active over here: https://github.com/Automattic/jetpack/pull/23871/files#:~:text=add_action(%20%27init%27%2C%20__NAMESPACE__,%7D? 

I'll keep thinking about what's the right way to do this as well.

jeherve commented:
That's a good question. While we may be able to shortcircuit things from within the package, I wonder if we should do that within the config package instead. This hasn't come up until now but will as we keep extracting features from modules.
@zinigor Since you worked on the Config package, I wonder what your take would be on this problem?

zinigor commented:
Ouch, sorry, I shouldn't have let this drag on for so long.

So, the way `ensure` was meant to work is that the plugin would sort of declare its need for this feature. You have the right idea to call `$config->ensure( 'publicize' )`, but this call only sets a flag that your plugin is interested in using it.

You are probably going to be better off with a two step approach.

1. Somewhere when you're loading your code you call `$config->ensure( 'publicize' )`
2. You also set a hook for a specific action that runs when Publicize is initialized: https://github.com/Automattic/jetpack/blob/master/projects/packages/config/src/class-config.php#L196. In this case it would be `jetpack_feature_publicize_enabled`.
3. In that hook you know that the module has been loaded already and is ready to use. 
4. If the hook doesn't run, it means that the module is not available in the current installation.

Hopefully that works for you, please let me know if that's not what you want.

p-jackson committed:
[not verified] Only force post thumbnail support if publicize is enabled

jeherve commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -219,8 +219,12 @@ public function __construct() {
 		add_action( 'wp_ajax_test_publicize_conns', array( $this, 'test_publicize_conns' ) );
 
-		add_action( 'init', array( $this, 'add_post_type_support' ) );
+		add_action( 'init', array( $this, 'add_post_type_support' ), 8 );
```
Do we need the custom priority for this one too?

p-jackson replied:
So we call `get_post_types_by_support()` inside the action handler in order to decide which post types to add `post-thumbnails` support to. However `get_post_types_by_support( 'publicize' )` doesn't return anything until `add_post_type_support()` is called. Since the action handler below gets called in priority 8, this one needs to as well.

I should have left a comment. I'll try to think of something succinct that explains that.

A lot of the messiness in this PR could be avoided if we didn't worry about making an easy way to undo the behaviour of this PR. But I think it's worth doing because we are forcing a theme to support something when the theme developer may not expect that.

jeherve commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1456,3 +1460,13 @@ function publicize_calypso_url() {
 	return Redirect::get_url( 'calypso-marketing-connections', array( 'site' => ( new Status() )->get_site_suffix() ) );
 }
+
+/**
+ * Adds support for the post-thumbnails feature, regardless of underlying theme support.
+ *
+ * This ensures the featured image UI appears in the editor, allowing the user to
+ * explicitly set an image for their social media post.
+ */
+function add_theme_post_thumbnails_support() {
+	add_theme_support( 'post-thumbnails', get_post_types_by_support( 'publicize' ) );
+}
```
Could this be in the class, instead of living outside of it?

p-jackson replied:
Discussed this one above https://github.com/Automattic/jetpack/pull/23871#discussion_r846856938
Making it a non-member function makes it easier for someone to use `remove_action` to remove this behaviour if that's not the way they want their theme to work. That was the idea anyway, maybe there's another way to allow people to override this hacking in of `post-thumbnails` support?

p-jackson committed:
[not verified] Comment to explain custom action priority

p-jackson commented:
Rebased now that #24451 has been merged. The explicit check that the module is enabled is still required because the `jetpack_feature_publicize_enabled` action is fired even when Publicized is disabled in the settings.

I've added a comment to explain why the custom action priorities are needed and tested everything's still working. What do you think @jeherve?

pablinos commented:
> The explicit check that the module is enabled is still required because the jetpack_feature_publicize_enabled

We only instantiate the class (and therefore load the code) if the module is enabled. I think if it's not loaded then `get_post_types_by_support( 'publicize' )` would be empty, so that could work as an implicit check if you didn't want to have an explicit module check.

p-jackson commented:
> We only instantiate the class (and therefore load the code) if the module is enabled.

This is what I'd expect but even after #24451 it still seems to be loaded all the time. `$config->ensure( 'publicize' )` is always called because of this code in [the main Jetpack plugin](https://github.com/Automattic/jetpack/blob/d20049ade86703517f345f973fe9d6d9bc6c8baa/projects/plugins/jetpack/class.jetpack.php#L826). That ultimately filters through to [this call to `Publicize_Setup::configure()`](https://github.com/Automattic/Jetpack/blob/48dd16ce512e23d0d8d0d28f6f8c27fd72aad30d/projects/packages/config/src/class-config.php#L263) which [installs a hook for the `on_jetpack_feature_publicize_enabled` action](https://github.com/Automattic/Jetpack/blob/48dd16ce512e23d0d8d0d28f6f8c27fd72aad30d/projects/packages/publicize/src/class-publicize-setup.php#L18) which initiates the class and loads the code.

I think maybe the misunderstanding is coming from the idea that the `on_jetpack_feature_publicize_enabled` action would only be fired when the feature is enabled in the settings, but [it's actually fired whenever the `$config->ensure()` method is called](https://github.com/Automattic/Jetpack/blob/48dd16ce512e23d0d8d0d28f6f8c27fd72aad30d/projects/packages/config/src/class-config.php#L199). And that call to `ensure( 'publicize' )` is hardcoded in the root Jetpack plugin file I linked above.

Is this expected? Or is there still a bit of a mix up about when the Publicize code should load @pablinos

pablinos commented:
> Or is there still a bit of a mix up about when the Publicize code should load @pablinos

Yes, essentially that's the problem I think! Weirdly I thought it was now conditional on the module being enabled. @spsiddarthan if we add the module check into `on_jetpack_feature_publicize_enabled` (or conditionally call `ensure` based on the module being enabled) will that cause the same problem we were seeing before?

spsiddarthan commented:
I think calling the ensure code conditionally based on whether the publicize module is active could be a solution, let me try it out!

spsiddarthan commented:
I am still working on fixing the init code logic. I have a few tests failing on my PR that I've to address.

spsiddarthan commented:
@p-jackson, the publicize init code refactoring is still in progress, I hope to finish it tomorrow, I need to adapt some failing tests. If you would like, I can take this fix over as well and get it merged once everything is working fine.

p-jackson commented:
@spsiddarthan Yes that's fine with me :) Although I'll still keep an eye on it because as even after the `( new Modules() )->is_active( 'publicize' )` check has been removed, there may still be some questions about the original fix that come up during review. But more than happy for you to update and rebase once your other PR is merged.

spsiddarthan commented:
@p-jackson, The PR to fix the publicize initialization has been merged, our team is a meetup, but I can take a look at getting this PR tested and merged tomorrow or the day after.

p-jackson committed:
Remove is_active check before forcing thumbnail support

The publicize class is no longer instantiated unless the feature is
enabled, so the extra is_active check is no longer necessary.

p-jackson commented:
@spsiddarthan I've just rebased, removed the explicit `is_active` check, retested, and it's working as expected!

jeherve committed:
Update dependencies
